### DISCUSSION TITLE
Anisotropic Diffusivity using Euler Angles

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#19613] [Q&A Modules: General] [answered]
Hello

I want to rotate the Anisotropic diffusivity tensor as per the crystallographic orientation of grains using Euler angles. In addition to that, I want to read the Euler angles from the file. So far I can able to simulate anisotropic diffusion by writing the kernels. Can you please help to find a way to rotate the anisotropic diffusivity tensor as per the grain orientation using Euler angles read from the file?

Kernel files and input files are as below:

```
INPUT FILE
[Mesh]
  type = GeneratedMesh # Can generate simple lines, rectangles and rectangular prisms
  dim = 2 # Dimension of the mesh
  nx = 20 # Number of elements in the x direction
  ny = 20 # Number of elements in the y direction
  xmax = 1.0 # Length of test chamber
  ymax = 1.0 # Test chamber radius
[]

[Variables]
  [diffused]
  []
[]



[Kernels]
  [diffusion]
    type = Concentration # Zero-gravity, divergence-free form of Darcys law
    variable = diffused # Operate on the "concentration" variable from above
    diffusivity = '7.08e-5 0 0
                  0    0     0
                  0    0     0'
  [../]

  [./euler]
    type = DiffusionTimeDerivative
    variable = diffused
  [../]
[]


[BCs]

 [./flux_bc]
    type = CoupledNeumannBC
    variable = diffused
    boundary = left
    Flux_val = -0.0001
     
  [../]

  [outlet]
    type = ADDirichletBC
    variable = diffused
    boundary = right
    value = 0  
  []
[]

[Executioner]
  type = Transient # Steady state problem
  solve_type = NEWTON # Perform a Newton solve
  num_steps = 250 # Run for 75 time steps, solving the system each step.
  dt = 1 # each time step will have duration "1"
  petsc_options_iname = '-pc_type -pc_hypre_type' # PETSc option pairs with values below
  petsc_options_value = ' hypre    boomeramg'
[]

[Outputs]
  execute_on = 'initial timestep_end'
  interval = 10
  exodus = true
[]





KERNELS

Concentration.h

#pragma once
#include "ADKernel.h"

class Concentration : public ADKernel
{
public:
  static InputParameters validParams();

  Concentration(const InputParameters & parameters);

protected:
  /// ADKernel objects must override precomputeQpResidual
  virtual ADReal computeQpResidual() override;

  /// References to be set from input file
  const RealTensorValue & _diffusivity;
};




Concentration.C

#include "Concentration.h"

registerMooseObject("BabblerApp", Concentration);

InputParameters
Concentration::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addClassDescription("Compute the concentration term for Fickian diffusion ($c$) equation: "
                             "$-\\nabla \\cdot \\frac{\\mathbf{K}}{\\mu} \\nabla p = 0$");

  params.addRequiredParam<RealTensorValue>("diffusivity", "The permeability ($\\mathrm{K}$) of the fluid.");

  return params;
}

Concentration::Concentration(const InputParameters & parameters)
  : ADKernel(parameters),
    _diffusivity(getParam<RealTensorValue>("diffusivity"))
{
}

ADReal
Concentration::computeQpResidual()
{
  return (_diffusivity) * _grad_test[_i][_qp] * _grad_u[_qp];
}
```

Thanks in advance. If possible please suggest some examples as well.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Not sure about examples, I ll let experts on these systems weigh in. 
I think if you need to rotate the diffusivity tensor based on the local grain orientation, that will have to be done in the MaterialProperty. 
Material properties, such as `_diffusivity` here, are re-computed every time a new element is being considered. 

PS there are existing utilities for rotating tensors if that can save you time

Guillaume

**Reply by avtarsinghh1991** [reply]
Thanks. Yes, I need to rotate it as per the orientation angle. D' = Q^T D Q, where Q is the orientation matrix. But I have a question about how to implement it in the above kernels. And I want to read the euler angles from the file.

There is one object called RotationTensor in the MOOSE but I cant able to figure out how to use that to rotate the diffusivity.

**Reply by GiudGiud** [reply]
I think the code to rotate tensor might be in utils/RankXXXTensor.h

I think it may be more natural to do the rotation in the computeQpProperties of the material rather than the kernel.
It's likely to work in the kernel too, just that if you have another object or another kernel that will depend on this, then you could forget to do the rotation there. 

Reading Euler angles from a file is done by an EulerAngleFileReader
https://mooseframework.inl.gov/source/userobjects/EulerAngleFileReader.html
which is then provided to other objects like for example a GrainTrackerElasticity
https://mooseframework.inl.gov/moose/source/postprocessors/GrainTrackerElasticity.html

You can have a look at the latter's object s code to see how Euler angles are being passed
https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/postprocessors/GrainTrackerElasticity.C
looks like it s just getEulerAngles called on the reader with the grain id as the parameter. 
The header (or the doxygen) for the EulerAngleFileReader will have all the options.

**Reply by avtarsinghh1991** [reply]
Thank you. As per your suggestion, I am trying to rotate the diffusivity using computeQpProperties of the material.

I am taking "ComputeElasticityTensorCP.h" and "ComputeElasticityTensorCP.C" as reference case.

https://mooseframework.inl.gov/docs/doxygen/modules/ComputeElasticityTensorCP_8h_source.html
https://mooseframework.inl.gov/docs/doxygen/modules/ComputeElasticityTensorCP_8C_source.html

However, when I am implementing the same, I am getting the following error

 error: 'MooseADWrapper<libMesh::TensorValue<double>, false>' {aka 'class libMesh::TensorValue<double>'} has no member named 'rotate'  _Diff_Tensor[_qp] = _Diff_Matrix[_qp].rotate(_crysrot[_qp]);



AnisoDiffusivity.h is as below:::

```
#pragma once

#include "Material.h"
#include "ElementPropertyReadFile.h"
#include "RankTwoTensor.h"
#include "RotationTensor.h"

class AnisoDiffusivity : public Material
{
public:
  static InputParameters validParams();
  
  AnisoDiffusivity(const InputParameters & parameters);

protected:
    virtual void computeQpProperties() override;

    virtual void assignEulerAngles();

    RealVectorValue _Euler_angles;

    const ElementPropertyReadFile * const _read_prop_user_object;

    MaterialProperty<RealVectorValue> & _Euler_angles_mat_prop;

    MaterialProperty<RankTwoTensor> & _crysrot;
     
    RotationTensor _R;

    MaterialProperty<Real> & _diffusivity;

    MaterialProperty<RealTensorValue> & _Diff_Matrix;
    MaterialProperty<RealTensorValue> & _Diff_Tensor;
};
```




AnisoDiffusivity.C as below:

```
#include "AnisoDiffusivity.h"
#include "RotationTensor.h"

registerMooseObject("BabblerApp", AnisoDiffusivity);

InputParameters
AnisoDiffusivity::validParams()
{
  InputParameters params = Material::validParams();
  params.addClassDescription("Compute an diffusion tensor for anisotropic transport.");
  params.addParam<UserObjectName>("read_prop_user_object",
                                      "The ElementReadPropertyFile"
                                      "GeneralUserObject to read element "
                                      "specific property values from file");
  params.addRequiredParam<RealTensorValue>("Diff_Matrix", "Anisotropic Diffusivity Tensor");
  return params;
}

AnisoDiffusivity::AnisoDiffusivity(const InputParameters & parameters)
  : Material(parameters),
   _read_prop_user_object(isParamValid("read_prop_user_object")
                                  ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                                   : nullptr),
    _Euler_angles_mat_prop(declareProperty<RealVectorValue>("Euler_angles")),
    _crysrot(declareProperty<RankTwoTensor>("crysrot")),
    _R(_Euler_angles),
    _diffusivity(declareProperty<Real>("diffusivity")),
    _Diff_Matrix(declareProperty<RealTensorValue>("Diff_Matrix")),
    _Diff_Tensor(declareProperty<RealTensorValue>("Diff_Tensor"))
{
}

 void
    AnisoDiffusivity::assignEulerAngles()
    {
      if (_read_prop_user_object)
      {
        _Euler_angles_mat_prop[_qp](0) = _read_prop_user_object->getData(_current_elem, 0);
        _Euler_angles_mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
        _Euler_angles_mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);
      }
      else
        _Euler_angles_mat_prop[_qp] = _Euler_angles;
    }


 void
    AnisoDiffusivity::computeQpProperties()
    {
     assignEulerAngles();
     
     _R.update(_Euler_angles_mat_prop[_qp]);
     
     _crysrot[_qp] = _R.transpose();

     _Diff_Tensor[_qp] = _R*_Diff_Matrix[_qp]*_R.transpose();
    }
```

On top of that Diff_Matrix values given in the input file is not properly passed to the kernels.

Can you please see what is the issue?  I shall be very thankful.

**Reply by hugary1995** [reply]
The error message is pretty clear about what's going on here: `error: 'MooseADWrapper<libMesh::TensorValue, false>' {aka 'class libMesh::TensorValue'} has no member named 'rotate'`. You should use `RankTwoTensor` instead of `RealTensorValue`.

**Reply by GiudGiud** [reply]
I agree for rotate. 

For diff_matrix not being passed, what are you doing is declaring a material property with the name passed as the diff_matrix parameter, with `declareProperty<RealTensorValue>("Diff_Matrix")`. You are not filling said matrix. 

To do that please have a look at the code for ADGenericConstantRankTwoTensor
https://mooseframework.inl.gov/source/materials/ADGenericConstantRankTwoTensor.html
https://github.com/idaholab/moose/blob/next/framework/src/materials/GenericConstantRankTwoTensor.C
this line in particular: `  _tensor.fillFromInputVector(getParam<std::vector<Real>>("tensor_values"));`

**Reply by avtarsinghh1991** [reply]
Thank you very much, everyone. Now I can able to read the diffusion matrix. I can also able to rotate the matrix as

 _Diff_Matrix[_qp].rotate(_crysrot[_qp]);

But When I am trying to assign this value to another RankTwoTensor as

_Temp[_qp] = _Diff_Matrix[_qp].rotate(_crysrot[_qp]);

the following error appears:

 error: no match for 'operator=' (operand types are 'MooseADWrapper<RankTwoTensorTempl<double>, false>' {aka 'RankTwoTensorTempl<double>'} and 'void')
   58 |      _Temp[_qp] = _Diff_Matrix[_qp].rotate(_crysrot[_qp]);




Header file is as


#pragma once

#include "Material.h"
#include "ElementPropertyReadFile.h"
#include "RankTwoTensor.h"
#include "RotationTensor.h"

class AnisoDiffusivity : public Material
{
public:
  static InputParameters validParams();
  
  AnisoDiffusivity(const InputParameters & parameters);

protected:
    virtual void computeQpProperties() override;

    virtual void assignEulerAngles();

    RealVectorValue _Euler_angles;

    const ElementPropertyReadFile * const _read_prop_user_object;

    MaterialProperty<RealVectorValue> & _Euler_angles_mat_prop;

    MaterialProperty<RankTwoTensor> & _crysrot;
    MaterialProperty<RankTwoTensor> & _Diff_Matrix;
    MaterialProperty<RankTwoTensor> & _Temp;
     
    RotationTensor _R;

    RankTwoTensor _diffusivity;

    MaterialProperty<RealTensorValue> & _Diff_Tensor;
};



Source file is as follows:

#include "AnisoDiffusivity.h"
#include "RotationTensor.h"
#include "GenericConstantRankTwoTensor.h"

registerMooseObject("BabblerApp", AnisoDiffusivity);

InputParameters
AnisoDiffusivity::validParams()
{
  InputParameters params = Material::validParams();
  params.addClassDescription("Compute an diffusion tensor for anisotropic transport.");
  params.addParam<UserObjectName>("read_prop_user_object",
                                      "The ElementReadPropertyFile"
                                      "GeneralUserObject to read element "
                                      "specific property values from file");
  params.addRequiredParam<std::vector<Real>>("diffusivity", "Stiffness tensor for material");
  return params;
}

AnisoDiffusivity::AnisoDiffusivity(const InputParameters & parameters)
  : Material(parameters),
   _read_prop_user_object(isParamValid("read_prop_user_object")
                                  ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                                   : nullptr),
    _Euler_angles_mat_prop(declareProperty<RealVectorValue>("Euler_angles")),
    _crysrot(declareProperty<RankTwoTensor>("crysrot")),
    _Diff_Matrix(declareProperty<RankTwoTensor>("Diff_Matrix")),
    _Temp(declareProperty<RankTwoTensor>("Temp")),
    _R(_Euler_angles),
    _Diff_Tensor(declareProperty<RealTensorValue>("Diff_Tensor"))
{
	_diffusivity.fillFromInputVector(getParam<std::vector<Real>>("diffusivity"));
}

 void
    AnisoDiffusivity::assignEulerAngles()
    {
      if (_read_prop_user_object)
      {
        _Euler_angles_mat_prop[_qp](0) = _read_prop_user_object->getData(_current_elem, 0);
        _Euler_angles_mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
        _Euler_angles_mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);
      }
      else
        _Euler_angles_mat_prop[_qp] = _Euler_angles;
    }


 void
    AnisoDiffusivity::computeQpProperties()
    {
     assignEulerAngles();
     
     _R.update(_Euler_angles_mat_prop[_qp]);
     
     _crysrot[_qp] = _R.transpose();

     _Temp[_qp] = _Diff_Matrix[_qp].rotate(_crysrot[_qp]);
     _Diff_Tensor[_qp] = _Diff_Matrix[_qp].rotate(_crysrot[_qp]);
    }

**Reply by GiudGiud** [reply]
Please format your code snippets with triple quotes. This makes them easier to read. 

Why do you need this _Temp material property?
Seems like you have everything you need in diff_matrix already

Try to separate the operations too
```
_Diff_Matrix[_qp].rotate(_crysrot[_qp]);
 _Temp[_qp] = _Diff_Matrix[_qp];
```

**Reply by avtarsinghh1991** [reply]
Thank you very much. I thought it need to be passed to a variable in order to use it in the source code.

Now, I implemented it as per your suggestion but didn't get the intended results. I am trying to print the _Diff_Matrix after the rotation

"""   _Diff_Matrix[_qp].rotate(_crysrot[_qp]); """

""" std::cout << _Diff_Matrix[_qp] << std::endl; """

and find that there is no effect of the rotation, whatever Euler angles I choose results are the same which should not be the case. I am getting the same results with and without the rotation. 

Result: Diff_Matrix is same before and after the rotation
(xx,xy,xz)=(7.08e-05,        0,        0)
(yx,yy,yz)=(       0, 7.08e-05,        0)
(zx,zy,zz)=(       0,        0, 7.08e-05)

Can you please suggest if there are any loose ends in the code?

**Reply by GiudGiud** [reply]
looks like you can do either, rotate in place (second) or create a rotated copy (first definition)
```
  /**
   * Returns a rotated version of the tensor data given a rank two tensor rotation tensor
   * _coords[i][j] = R_ij * R_jl * _coords[k][l]
   * @param R rotation matrix as another RankTwoTensorTempl
   */
  RankTwoTensorTempl<T> rotated(const RankTwoTensorTempl<T> & R) const;

  /**
   * rotates the tensor data given a rank two tensor rotation tensor
   * _coords[i][j] = R_ij * R_jl * _coords[k][l]
   * @param R rotation matrix as a RankTwoTensorTempl
   */
  void rotate(const RankTwoTensorTempl<T> & R);
```

**Reply by avtarsinghh1991** [reply]
Thank you. Actually, I am very new to MOOSE. Can you please let me know how to implement them in my case? 

```
#include "AnisoDiffusivity.h"
#include "RotationTensor.h"
#include "GenericConstantRankTwoTensor.h"
#include "RankTwoTensor.h"

registerMooseObject("BabblerApp", AnisoDiffusivity);

InputParameters
AnisoDiffusivity::validParams()
{
  InputParameters params = Material::validParams();
  params.addClassDescription("Compute an diffusion tensor for anisotropic transport.");
  params.addParam<UserObjectName>("read_prop_user_object",
                                      "The ElementReadPropertyFile"
                                      "GeneralUserObject to read element "
                                      "specific property values from file");
  params.addRequiredParam<std::vector<Real>>("diffusivity", "Stiffness tensor for material");
  return params;
}

AnisoDiffusivity::AnisoDiffusivity(const InputParameters & parameters)
  : Material(parameters),
   _read_prop_user_object(isParamValid("read_prop_user_object")
                                  ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                                   : nullptr),
    _Euler_angles_mat_prop(declareProperty<RealVectorValue>("Euler_angles")),
    _crysrot(declareProperty<RankTwoTensor>("crysrot")),
    _Diff_Matrix(declareProperty<RankTwoTensor>("Diff_Matrix")),
    _R(_Euler_angles),
    _Diff_Tensor(declareProperty<RealTensorValue>("Diff_Tensor"))
{
    _diffusivity.fillFromInputVector(getParam<std::vector<Real>>("diffusivity"));
}

 void
    AnisoDiffusivity::assignEulerAngles()
    {
      if (_read_prop_user_object)
      {
        _Euler_angles_mat_prop[_qp](0) = _read_prop_user_object->getData(_current_elem, 0);
        _Euler_angles_mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
        _Euler_angles_mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);
      }
      else
        _Euler_angles_mat_prop[_qp] = _Euler_angles;
    }

 void
    AnisoDiffusivity::computeQpProperties()
    {
     assignEulerAngles();
     
     _R.update(_Euler_angles_mat_prop[_qp]);
     
     _crysrot[_qp] = _R.transpose();

     _Diff_Matrix[_qp] = _diffusivity; 
     
     _Diff_Matrix[_qp].rotate(_crysrot[_qp]);
    
     std::cout << _Diff_Matrix[_qp] << std::endl;
    }
```

**Reply by avtarsinghh1991** [reply]
Thank you very much for your help. It started working for a single block or domain. However, I still can't able to run it for multiple blocks. I made a abaqus mesh (.inp) file with four blocks as shown below:

*ELEMENT, TYPE=S4R, ELSET=EB3
*ELEMENT, TYPE=S4R, ELSET=EB4
*ELEMENT, TYPE=S4R, ELSET=EB1
*ELEMENT, TYPE=S4R, ELSET=EB2



**and USEROBJECT defined in the input file is as follows:**

[UserObjects]
  [./euler_angle_read]
    type = ElementPropertyReadFile
    prop_file_name = 'input_file.txt'
    nprop = 3
    read_type = block
    nblock = 4
  [../]
[]


**And when I am running, the following error appears:**

* ERROR *
/home/asingh/projects/babbler/problems/reading_data.i:40: (UserObjects/euler_angle_read/nblock):
    Element block id 4 greater than than total number of blocks in mesh 4

Update: If I am putting nblock = 7 or greater, the code is working but the results are wrong. I am using same mesh file with four blocks in some other moose application and it is working absolutely fine.

**Update: Printed data from moose**

Elems:
    Total:                 1600
    Local:                 400
    Min/Max/Avg:           400/400/400
  Num Subdomains:          7
  Num Partitions:          4
  Partitioner:             metis

How the moose is reading 7 subdomains when there are only 4 in the mesh file?

Please help.
Thanks

**Reply by GiudGiud** [reply]
Hello

That is exactly the problem. Can you run with `--mesh-only` on the command line and examine the mesh MOOSE sees?
The Euler Angle reader is very sensitive to this I think, it needs to see the right number of blocks and they need to be numbered contiguously iirc (I should double check that last claim). 

Guillaume

**Reply by avtarsinghh1991** [reply]
Hello Guillaume

Thank you very much. Yes. I run it and got the following results

For your reference: Mesh domain is 4 square connecting each other.


 Mesh Information:
  elem_dimensions()={2}
  spatial_dimension()=2
  n_nodes()=1681
    n_local_nodes()=447
  n_elem()=1600
    n_local_elem()=400
    n_active_elem()=1600
  n_subdomains()=7
  n_partitions()=4
  n_processors()=4
  n_threads()=1
  processor_id()=0
  is_prepared()=true
  is_replicated()=true

 Mesh Bounding Box:
  Minimum: (x,y,z)=(       0,        0,        0)
  Maximum: (x,y,z)=(       1,        1,        0)
  Delta:   (x,y,z)=(       1,        1,        0)

 Mesh Element Type(s):
  QUAD4

 Mesh Nodesets:
  Nodeset 0 (SS1), 1681 nodes
   Bounding box minimum: (x,y,z)=(       0,        0,        0)
   Bounding box maximum: (x,y,z)=(0.400002,     0.55,        0)
   Bounding box delta: (x,y,z)=(0.400002,     0.55,        0)

 Mesh Sidesets:
  Sideset 0 (SS1), 80 sides (EDGE2), 80 elems (QUAD4), 84 nodes
   Side volume: 2
   Bounding box minimum: (x,y,z)=(       0,        0,        0)
   Bounding box maximum: (x,y,z)=(       1,        1,        0)
   Bounding box delta: (x,y,z)=(       1,        1,        0)

 Mesh Edgesets:
  None

 Mesh Subdomains:
  Subdomain 0 (EB1_QUAD4): 380 elems (QUAD4, 380 active), 420 active nodes
   Volume: 0.2375
   Bounding box minimum: (x,y,z)=(0.0249974,        0,        0)
   Bounding box maximum: (x,y,z)=(     0.5,      0.5,        0)
   Bounding box delta: (x,y,z)=(0.475003,      0.5,        0)
  Subdomain 1 (EB2_QUAD4): 380 elems (QUAD4, 380 active), 420 active nodes
   Volume: 0.237501
   Bounding box minimum: (x,y,z)=(     0.5, 0.0249973,        0)
   Bounding box maximum: (x,y,z)=(       1,      0.5,        0)
   Bounding box delta: (x,y,z)=(     0.5, 0.475003,        0)
  Subdomain 2 (EB3_QUAD4): 380 elems (QUAD4, 380 active), 420 active nodes
   Volume: 0.2375
   Bounding box minimum: (x,y,z)=(       0,      0.5,        0)
   Bounding box maximum: (x,y,z)=(     0.5, 0.975003,        0)
   Bounding box delta: (x,y,z)=(     0.5, 0.475003,        0)
  Subdomain 3 (EB4_QUAD4): 380 elems (QUAD4, 380 active), 420 active nodes
   Volume: 0.2375
   Bounding box minimum: (x,y,z)=(     0.5,      0.5,        0)
   Bounding box maximum: (x,y,z)=(0.975003,        1,        0)
   Bounding box delta: (x,y,z)=(0.475003,      0.5,        0)
  Subdomain 4 (SS1_E1_QUAD4): 74 elems (QUAD4, 74 active), 156 active nodes
   Volume: 0.0462486
   Bounding box minimum: (x,y,z)=(       0,        0,        0)
   Bounding box maximum: (x,y,z)=(       1,        1,        0)
   Bounding box delta: (x,y,z)=(       1,        1,        0)
  Subdomain 5 (SS1_E2_QUAD4): 3 elems (QUAD4, 3 active), 12 active nodes
   Volume: 0.00187489
   Bounding box minimum: (x,y,z)=(       0,    0.475,        0)
   Bounding box maximum: (x,y,z)=(       1,        1,        0)
   Bounding box delta: (x,y,z)=(       1,    0.525,        0)
  Subdomain 6 (SS1_E4_QUAD4): 3 elems (QUAD4, 3 active), 12 active nodes
   Volume: 0.00187488
   Bounding box minimum: (x,y,z)=(       0,        0,        0)
   Bounding box maximum: (x,y,z)=(       1,        1,        0)
   Bounding box delta: (x,y,z)=(       1,        1,        0)
  Global mesh volume = 1


As far as I understand, MOOSE is considering SIDESETS as domain.

**Reply by GiudGiud** [reply]
It really should not. You can rename the blocks to merge them with this mesh generator:
https://mooseframework.inl.gov/source/meshgenerators/RenameBlockGenerator.html

I would really zoom in on these regions in paraview to check that the elements are not weird.

**Reply by avtarsinghh1991** [reply]
Hey Guillaume.

Thank you very much. I am testing it on the four blocks domain so that I can understand the system before running it for a number of grains.

It's just a simple mesh file with square elements. When I am exporting it as a .inp file, MOOSE is taking side sets as a domain. However, when I have exported the mesh as exodus (.e) type and then rename the block from 1 2 3 4 to 0 1 2 3, the code started working absolutely fine.
I tried with different geometry but the same issue appears.
I am just wondering is there anything to do with the subroutine I wrote for anisotropic diffusion or reading mesh is independent of it?

Please help.
Thanks

**Reply by GiudGiud** [reply]
Hello

Glad you found a working solution. 
The mesh generation code is independent from the material & kernels you worked on. 
I ll keep your issues in mind and see if I can reproduce them on an inp mesh later.

Guillaume

**Reply by avtarsinghh1991** [reply]
Thanks. I am looking forward to hearing from you. 

Just one more thing. Do you know any editor or method so that I can read the exodus file ?

Best
Avtar

**Reply by GiudGiud** [reply]
we all use paraview pretty much to look at exodus files. 
5.9 works well, there's an issue for us in a later version, it should be fixed soon

**Reply by avtarsinghh1991** [reply]
Thanks. I mean to read the content, the exodus is a binary file so can't able to see anything except some symbols when open in gvim or notepad?

**Reply by GiudGiud** [reply]
Yeah I wouldnt use those. If you want to be writing or reading exodus files, you can do it with C++ code. 
There's some examples and a lot of utilities already existing in moose
https://github.com/idaholab/moose/blob/next/framework/src/outputs/Exodus.C

**Comment by dschwen** [comment]
The right approach is to derive another object from `GrainDataTracker` that is similar to `GrainTrackerElasticity`, but provides a rotated `RealTensorValue`. I developed `GrainDataTracker` to deal with grain tracker edge cases like grain splitting, where new grain numbers can emerge that you won't have data for in your Euler angle table.

You can internally use `RankTwoTensor`s to perform the matrix rotation and then downcast to a `RealTensorValue`.

**Reply by avtarsinghh1991** [reply]
Thank you very much. I tried that it works to a certain extent but I can't able to assign the rotated value to a certain variable. I am getting the error which I listed in the above reply.
